src/core/ddsi/src/ddsi_typelookup.c

```diff
@@ -0,0 +1,356 @@
+  ddsrt_free (templ.type_id);
+  return tlm;
+}
+
+struct tl_meta * ddsi_tl_meta_lookup (struct ddsi_domaingv *gv, const type_identifier_t *type_id)
+{
+  struct tl_meta *tlm;
+  ddsrt_mutex_lock (&gv->tl_admin_lock);
+  tlm = ddsi_tl_meta_lookup_locked (gv, type_id);
+  ddsrt_mutex_unlock (&gv->tl_admin_lock);
+  return tlm;
+}
+
+
+static bool guid_prefix_empty (const ddsi_guid_prefix_t *prefix)
+{
+  return prefix->u[0] == 0 && prefix->u[1] == 0 && prefix->u[2] == 0;
+}
+
+void ddsi_tl_meta_ref (struct ddsi_domaingv *gv, const type_identifier_t *type_id, const struct ddsi_sertype *type, const ddsi_guid_t *src_ep_guid, const ddsi_guid_prefix_t *dst)
+{
+  assert (type_id || type);
+  GVTRACE ("ref tl_meta");
+  type_identifier_t *tid;
+  if (type_id != NULL)
+    tid = (type_identifier_t *) type_id;
+  else
+  {
+    GVTRACE (" sertype %p", type);
+    tid = ddsi_typeid_from_sertype (type);
+  }
+  GVTRACE (" tid "PTYPEIDFMT, PTYPEID (*tid));
+
+  ddsrt_mutex_lock (&gv->tl_admin_lock);
+  struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, tid);
+  if (tlm == NULL)
+  {
+    tlm = ddsrt_malloc (sizeof (*tlm));
+    memset (tlm, 0, sizeof (*tlm));
+    tlm->gv = gv;
+    tlm->state = TL_META_NEW;
+    tlm->type_id = ddsi_typeid_dup (tid);
+    ddsrt_hh_add (gv->tl_admin, tlm);
+    GVTRACE (" new %p", tlm);
+  }
+  if (src_ep_guid != NULL)
+  {
+    bool found = false;
+    for (uint32_t n = 0; !found && n < tlm->n_endpoints; n++)
+    {
+      if (!memcmp (&tlm->endpoints[n], src_ep_guid, sizeof (*src_ep_guid)))
+        found = true;
+    }
+    if (!found)
+    {
+      tlm->n_endpoints++;
+      tlm->endpoints = ddsrt_realloc (tlm->endpoints, tlm->n_endpoints * sizeof (*tlm->endpoints));
+      memcpy (&tlm->endpoints[tlm->n_endpoints - 1], src_ep_guid, sizeof (*src_ep_guid));
+      GVTRACE (" add ep "PGUIDFMT, PGUID (*src_ep_guid));
+    }
+  }
+  if (tlm->sertype == NULL && type != NULL)
+  {
+    tlm->sertype = ddsi_sertype_ref (type);
+    tlm->state = TL_META_RESOLVED;
+    GVTRACE (" resolved");
+  }
+  if (dst != NULL && guid_prefix_empty (&tlm->dst_prefix) && !guid_prefix_empty (dst))
+    memcpy (&tlm->dst_prefix, dst, sizeof (*dst));
+  tlm->refc++;
+  GVTRACE (" state %d refc %u\n", tlm->state, tlm->refc);
+  ddsrt_mutex_unlock (&gv->tl_admin_lock);
+  if (type_id == NULL)
+    ddsrt_free (tid);
+}
+
+void ddsi_tl_meta_unref (struct ddsi_domaingv *gv, const type_identifier_t *type_id, const ddsi_guid_t *ep_guid)
+{
+  ddsrt_mutex_lock (&gv->tl_admin_lock);
+  struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, type_id);
+  assert (tlm != NULL);
+  ddsi_tl_meta_unref_impl_locked (gv, tlm, ep_guid);
+  ddsrt_mutex_unlock (&gv->tl_admin_lock);
+}
+
+static struct participant *get_typelookup_pp (const struct ddsi_domaingv *gv, const ddsi_guid_prefix_t *guid_prefix)
+{
+  struct entidx_enum_participant est;
+  struct participant *pp = NULL;
+  bool pp_found = false;
+  entidx_enum_participant_init (&est, gv->entity_index);
```

Comment:
There is no need to scan the participants:
* if `guid_prefix_empty(guid_prefix)`, one just picks the first (same code, minus the loop).
* otherwise, one can use the fact that all participants have entity id 0x1c1 (there is a name for it, I can't be bothered to look it up right now), and so one can just do a GUID lookup of (effectively) `{guid_prefix,0x1c1}`.